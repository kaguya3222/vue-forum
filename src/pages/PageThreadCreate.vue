<template>
  <div v-if="asyncDataStatus_ready" class="col-full push-top">
    <h1>
      Create new thread in <i>{{ forum.name }}</i>
    </h1>
    <thread-editor @save="save" @cancel="cancel" />
  </div>
</template>

<script>
import ThreadEditor from "../components/ThreadEditor";
import { mapGetters, mapActions } from "vuex";
import asyncDataStatus from "../mixins/asyncDataStatus";

export default {
  name: "PageThreadCreate",
  components: {
    "thread-editor": ThreadEditor
  },
  mixins: [asyncDataStatus],
  props: {
    forumId: {
      required: true,
      type: String
    }
  },
  computed: {
    ...mapGetters({
      forums: "forums/items"
    }),
    forum() {
      return this.forums[this.forumId];
    }
  },
  methods: {
    ...mapActions("forums", ["fetchForum"]),
    ...mapActions("threads", ["createThread"]),
    save({ text, title }) {
      this.createThread({
        forumId: this.forum[".key"],
        title: title,
        text: text
      }).then(thread => {
        this.$router.push({
          name: "ThreadShow",
          params: { threadId: thread[".key"] }
        });
      });
    },
    cancel() {
      const forumId = this.forumId;
      this.$router.push({ name: "Forum", params: { id: forumId } });
    }
  },
  created() {
    this.fetchForum({ id: this.forumId }).then(() => {
      this.asyncDataStatus_fetched();
    });
  }
};
</script>

<style scoped></style>
