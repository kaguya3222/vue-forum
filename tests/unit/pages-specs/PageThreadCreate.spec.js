import { createLocalVue, mount } from "@vue/test-utils";
import PageThreadCreate from "../../../src/pages/PageThreadCreate";
import rootStore from "@/store/";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("PageThreadCreate", function() {
  const store = new Vuex.Store({
    state: rootStore.state
  });
  store.dispatch = jest.fn();
  const $router = {
    push: jest.fn()
  };
  const wrapper = mount(PageThreadCreate, {
    propsData: {
      forumId: "-KpOx5Y4AqRr3sB4Ybwj"
    },
    mocks: {
      $router
    },
    localVue,
    store
  });
  test("Correctly shows thread creation form", () => {
    expect(wrapper).toMatchSnapshot();
  });
  describe("save", () => {
    const threadTitleInput = wrapper.find("#thread_title");
    const threadContentInput = wrapper.find("#thread_content");
    const threadCreationForm = wrapper.find("form");
    beforeEach(() => {
      threadTitleInput.element.value = "Thread title";
      threadTitleInput.trigger("input");
      threadContentInput.element.value = "Thread content";
      threadContentInput.trigger("input");
      threadCreationForm.trigger("submit");
    });
    test("Dispatches 'createThread' action after form is submitted", () => {
      expect(store.dispatch).toHaveBeenCalledWith("createThread", {
        forumId: wrapper.vm.forum[".key"],
        title: wrapper.vm.title,
        text: wrapper.vm.text
      });
    });
    test("Redirects user to 'ThreadShow' route after form is submitted", () => {
      expect($router.push).toHaveBeenCalledWith({
        name: "ThreadShow",
        params: expect.any(Object)
      });
    });
  });
  describe("cancel", function() {
    const cancelButton = wrapper.find(".btn-ghost");
    test("Redirects user to forum page when 'cancel' button is clicked", () => {
      cancelButton.trigger("click");
      expect($router.push).toHaveBeenCalledWith({
        name: "Forum",
        params: {
          id: wrapper.vm.forumId
        }
      });
    });
  });
});
