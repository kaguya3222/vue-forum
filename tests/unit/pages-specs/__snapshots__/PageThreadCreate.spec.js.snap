// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageThreadCreate Correctly shows thread creation form 1`] = `
<div class="col-full push-top">
  <h1>
    Create new thread in <i>Announcements</i></h1>
  <thread-editor-stub text="" title=""></thread-editor-stub>
</div>
`;
